### PR TITLE
Sqlite support

### PR DESCRIPTION
**Pull request by SquattingSocrates**
Extended the excellent work of @jtenner and built rust guest bindings against this branch. I'd keep this open for now because I want to build a larger interface for better integration with ORMs etc.

It's usable for testing though. If you add this FFI in the lunatic-rs lib:
```rust
pub mod sqlite {
    #[link(wasm_import_module = "lunatic::sqlite")]
    extern "C" {
        pub fn open(path: *const u8, path_len: usize, conn_id: *mut u32) -> u64;
        pub fn query_prepare(
            conn_id: u64,
            query_str: *const u8,
            query_str_len: u32,
            len_ptr: *mut u32,
            resource_id: *mut u32,
        ) -> ();
        pub fn query_result_get(resource_id: u64, write_buf: *const u8, write_buf_len: u32) -> ();
        pub fn drop_query_result(resource_id: u64) -> ();
        pub fn execute(conn_id: u64, exec_str: *const u8, exec_str_len: u32) -> u32;
    }
}

```

Then add a couple of convenience wrapper methods for the host functions:
```rust
use crate::host::api;
use std::path::Path;

pub fn open(path: &Path) -> u64 {
    let conn_id = 0u64;
    let path_str = path.to_str().unwrap();
    unsafe {
        api::sqlite::open(path_str.as_ptr(), path_str.len(), &mut (conn_id as u32));
    }
    conn_id
}
/// returns a tuple consisting of the length of data written to the buf
/// as well as the resource id
pub fn query_prepare(conn_id: u64, query: &str) -> (u32, u64) {
    let mut len = 0u32;
    let resource_id = 0u64;
    unsafe {
        api::sqlite::query_prepare(
            conn_id,
            query.as_ptr(),
            query.len() as u32,
            &mut len,
            &mut (resource_id as u32),
        );
        (len, resource_id)
    }
}
pub fn query_result_get(resource_id: u64, buf: &mut [u8]) -> () {
    unsafe {
        api::sqlite::query_result_get(resource_id, buf.as_ptr(), buf.len() as u32);
    }
}
pub fn drop_query_result(resource_id: u64) -> () {
    unsafe {
        api::sqlite::drop_query_result(resource_id);
    }
}
pub fn execute(conn_id: u64, exec_str: &str) -> u32 {
    unsafe { api::sqlite::execute(conn_id, exec_str.as_ptr(), exec_str.len() as u32) }
}
```

You'll be able to run this example:
```rust
use lunatic::sqlite;
use std::path::Path;

fn main() {
    println!("Opening connection to sqlite...");
    let db = sqlite::open(Path::new("./test_db"));
    println!("Connected to sqlite");
    let x = sqlite::execute(db, "CREATE TABLE IF NOT EXISTS movie(title, year, score)");
    println!("RESULT OF EXEC {:?}", x);

    sqlite::execute(
        db,
        r"
    INSERT INTO movie VALUES
        ('Monty Python and the Holy Grail', 1975, 8.2),
        ('And Now for Something Completely Different', 1971, 7.5)",
    );

    let (len, result_id) = sqlite::query_prepare(db, "SELECT title, score FROM movie");
    let mut buf = vec![0u8; len as usize];
    sqlite::query_result_get(result_id.into(), &mut buf);
    println!("RECEIVED DATA FROM SQLITE {:?}", String::from_utf8(buf));
}
```

### CODE REVIEW

File: crates/lunatic-common-api/src/lib.rs [hunk 1]
```diff
@@ -11,6 +11,37 @@ pub fn get_memory<T>(caller: &mut Caller<T>) -> std::result::Result<Memory, Trap
...
 
+// Call guest to allocate a Vec of size `size`
+pub fn allocate_guest_memory<'a, T: Send>(
+    caller: &'a mut Caller<T>,
+    size: u32,
+) -> Pin<Box<dyn Future<Output = Result<u32, Trap>> + Send + 'a>> {
+    Box::pin(async move {
+        let mut results = [Val::I32(0)];
+        let result = caller
+            .get_export("alloc")
```
Comment (jtenner) [inline crates/lunatic-common-api/src/lib.rs:22]: This is probably not safe for AssemblyScript users. This would also probably require the runtime to be exported, which is feasable, but I'm not sure what a good solution is here.

Comment (bkolobara) [reply crates/lunatic-common-api/src/lib.rs:22]: @jtenner Would manually exporting a function `alloc` that just calls `malloc` (or the AS equivalent) work? If you can allocate memory manually there is no reason why you couldn't just wrap it into an exported function with a well defined name.

Comment (SquattingSocrates) [reply crates/lunatic-common-api/src/lib.rs]: @jtenner I added the ability to set a different "allocator" function (called `set_custom_guest_allocator`) which will allow you set to set custom allocators for each connection. All you'd need to do is provide the allocator function and set it every time when creating a new sqlite connection.

**Review by bkolobara — COMMENTED** [review]
There is a lot of stuff here and I will need to get to some of this back after I sleep a bit on it. But here are some of my first observations :)

### Unrelated changes

There are some changes in here that are not related to SQLite. I double checked them, but they don't seem to be doing anything. They allocate some vectors and add `Clone` derives to some structs, but are never used. I can only assume they are leftovers from some previous attempts. This PR is already big and this makes it extra hard to understand everything. Will require some cleaning up.

### Memory allocation in the guest

This is the first API that introduces memory allocation from host inside the guest. I think this is a useful feature and should be made global (not related to SQLite). So that we can always return values to the guest without knowing the exact size beforehand.

Unfortunately, this implementation is too specific by depending on a SQLite connection to work. I really don't think we need to introduce extra complexity here by allowing per-connection renaming of the guest allocator. I can't think of any benefit of doing this. My preference would be a well known allocator exported from guest (called "lunatic_alloc") and maybe a way of dropping the memory from host ("lunatic_free"?), even we don't have a use case for it yet.

### Safety is hard to reason about

This is currently my biggest worry. It's almost impossible for me to tell if changes in this PR are safe. Mostly because it uses [a fork of the SQLite crate](https://github.com/SquattingSocrates/sqlite) that declares the `Statement` struct `Send` and `Sync` with the unsafe keyword. It also introduces some other changes. If we store this struct inside of a process state then they **must** be `Send` or it could lead to some really weird issues where the scheduler moves the statement to another thread while it's being changed. This would also require us to maintain a fork of a fairly big project. We have done this in the beginning with `Wasmer` and `Wasmtime` and it was a nightmare.

I also believe that this approach gives the guest too many ways to escape our sandbox because you can open **any** SQLite file on the filesystem, even if your process doesn't have rights to access any folders. The strings are just forwarded to the SQLite open command. I'm not sure what the best way of safeguarding this would be. Should we just respect the file/folder permissions of the process or completely abstract away SQLite stuff? For example, the guest opens "database_name" and we create a backing file in a well known location, but don't let them observe the file.

### Concurrency issues

How do processes share connections? Usually one process will open a connection and forward it to others, but would this be safe? Are connections thread safe? What if multiple processes open the same db and concurrently try to write something inside it?

I believe that we require much more structure here. Identify some common patterns of usage and expose APIs that are hard to be used wrongly.

### Making things more SQLite agnostic

Another thing I'm currently thinking about is that maybe we shouldn't bind this API too closely to SQLite, but let it be just a way of executing SQL and use SQLite as the current implementation. In case of distributed lunatic we would move to a distributed SQL implementation transparently. But I don't know if such a design is feasible?

File: crates/lunatic-common-api/src/lib.rs [hunk 2]
```diff
@@ -11,6 +11,27 @@ pub fn get_memory<T>(caller: &mut Caller<T>) -> Result<Memory> {
         .or_trap("Export `memory` is not a memory")
 }
 
+// Call guest to allocate a Vec of size `size`
+pub fn allocate_guest_memory<'a, T: Send>(
```
[the remark this answers is not in the document]
Comment (SquattingSocrates) [reply crates/lunatic-common-api/src/lib.rs:15]: Maybe we should use a universal code for "out of memory" errors? For calls that will allocate guest memory we could then require the guest to send an opaque pointer to which we write the error in case the process ran out of memory. On the other hand, maybe we should just let it crash? Depends on whether the process developer wants to have a graceful shutdown or not in such a case. I would personally like to crash the process and try to figure out what data I've been reading that caused memory to run out.

File: crates/lunatic-sqlite-api/src/sqlite_bindings.rs [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+        .or_trap("lunatic::sqlite::write_to_guest_vec")?;
+
+    alloc_vec
+        .write(&encoded_vec)
+        .or_trap("lunatic::sqlite::write_to_guest_vec")?;
+
+    // shift len to the left in order to send length and pointer
+    // back to the guest
+    let alloc_len = (encoded_vec.len() as u64) << 32;
+    Ok(alloc_ptr as u64 | alloc_len)
```
[the remark this answers is not in the document]
Comment (SquattingSocrates) [reply crates/lunatic-sqlite-api/src/sqlite_bindings.rs:457]: Alright, I'll change it to writing to a pointer then. However I do like returning just a single value 🙂

File: crates/lunatic-sqlite-api/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+pub const SQLITE_EMPTY: u32 = 16;
+pub const SQLITE_SCHEMA: u32 = 17;
+pub const SQLITE_TOOBIG: u32 = 18;
+pub const SQLITE_CONSTRAINT: u32 = 19;
+pub const SQLITE_MISMATCH: u32 = 20;
+pub const SQLITE_MISUSE: u32 = 21;
+pub const SQLITE_NOLFS: u32 = 22;
+pub const SQLITE_AUTH: u32 = 23;
+pub const SQLITE_FORMAT: u32 = 24;
+pub const SQLITE_RANGE: u32 = 25;
```
[the remark this answers is not in the document]
Comment (SquattingSocrates) [reply crates/lunatic-sqlite-api/src/lib.rs:185]: We could move this to `lunatic-sql` as part of the sqlite diesel bindings because that's ultimately who's using these constants. In the lunatic-sqlite-api crate we actually only need 2 constant - SQLITE_DONE and SQLITE_ROW. The rest is used by the ORM. Since breaking API changes would have to be handled properly in the ORM bindings anyway I think we can safely move them there. Especially if we want to create a more database-agnostic interface

**Comment by SquattingSocrates** [comment]
First of all, thank you for the feedback! There were indeed some leftover changes from multiple previous approaches, I have removed the ones you pointed out so far.
As for the other points:

> Memory allocation in the guest

I agree that allocation should probably be not specific to SQLite and made global. The issue with that is of course that you have no way of using guest allocations from multiple languages at the same time. Maybe if we add the ability to set the allocator per process_id? That will make this feature more flexible, while not giving up using multiple languages.
`lunatic_free` sounds like a good idea, even though I also don't see a use-case for it right now.

> Safety is hard to reason about

It would require a bit more testing of multiple lunatic processes reading and writing to the same connection, however that's practically not easy because a `SqliteConnection` is not Serialisable and cannot be sent to another process at the moment because certain underlying diesel structures not serialisable. So I don't even know how to test this, maybe just spawning many processes with their own clients?
It's also possible to use a different crate (libsys_sqlite) that is actually just a C ffi and with that we could probably use the original crate and not a fork. Either way I want to try to request a PR for the forked crate changes and hopefully it will be accepted.

Checking path rights is definitely a good idea, will add this. I think that one of the valuable things about sqlite is that the db is just a single easy-to-use `*.db` file and therefore the user should probably be able to retrieve it if they want that.


> Concurrency issues

At the moment processes don't share connections because the connection is not serialisable. However the db itself should be handling concurrent requests pretty well. There are also 3 different modes of concurrency for sqlite, but I would need to test this out a bit: https://www.sqlite.org/threadsafe.html.


> Making things more SQLite agnostic

Now that I made sqlite work with some [examples](https://github.com/SquattingSocrates/sqlite-lunatic-diesel-example) and looked at how the backend+connection for mysql and Postgres are implemented I think that with the current api we could add support for the diesel ORM with other db's as well. I'm not so sure about other ORMs or tools. If we're willing to put mysql and Postgres clients in the host (probably behind some feature) we could fairly well reuse the existing api for general sql execution. I don't have a great way to do pooling in the guest yet, especially if we need to "mirror" the pool from the host to the guest and would probably work on this once SQLite is a bit more tested as part of the vm.

**Comment by SquattingSocrates** [comment]
> Why not? What is the actual issue here that makes it necessary to dynamically name exported allocator function?

Because the function has to live in the guest somewhere, at least that's my understanding. This implies that it needs to be compiled to wasm and to some degree accessible from the guest code that's making the call to the host. I just don't know how to make the wasm function accessible to say AssemblyScript besides compiling the allocator separately and then include as "library" or wasm module. Maybe I got something wrong here, but that's how I understood it. Even in Rust with `lunatic-sql` and `lunatic-rs` (which have no direct relationship with each other) there would need to be a single allocator function provided somewhere externally which we could just include in the respective Cargo.toml as dependency. So with Rust it's pretty ok to do, but plugging that into other languages would be hard I guess.

> Dropping to the lowest level and giving up safe wrappers certainly solves the problem of forking the a higher level library and reducing the safety constraints

Yes, that's definitely the way to go 😄 
I'll stage some tests with multiple processes and test how multi-threading actually works. But as far as I understand statements, they're just pointers to an object that holds some bytecode from parsing the provided sql. I think the statement exists only on the connection on which it was created (might be wrong though), because in order to bind or step over the statement you need the connection. Actually, since I need to store the connection and the statements separately I changed the structure of `Statement` in my fork of the sqlite crate. That's probably the largest reason for forking it 🙂 

But in general, SQLite can have many open connections to multiple different databases (e.g. files on disk) and handle concurrent connections pretty well. SQLite is probably not the best out-of-the-box for countless processes spamming the db with writes while also doing a ton of reads at the same time, but given that most of db access' are reads it should work well for quite some time. There's also WAL-mode which should give us better performance

**Comment by bkolobara** [comment]
Couldn't we just export the allocator lets say in the `lunatic-rs` library as:
```rust
#[export_name = "lunatic_alloc"]
pub fn alloc(len: usize) -> *mut u8 {
    let mut buf = Vec::with_capacity(len);
    let ptr = buf.as_mut_ptr();
    std::mem::forget(buf);
    ptr
}
```

And do the same equivalent in `as-lunatic`? So every time someone uses the libraries they don't need to do anything special and it just works out of the box, because the exports already exist. I mean, they also need to do this with your approach. I'm just confused with the renaming and still can't really wrap my head around it. Just looks like an extra step that doesn't contribute much.

**Comment by SquattingSocrates** [comment]
> Couldn't we just export the allocator lets say in the `lunatic-rs` library as ... And do the same equivalent in `as-lunatic`? 

I would like this approach as well, but the issue with the sql crate for example is that it doesn't have a dependency on lunatic-rs and this way we'd avoid this dependency. Same goes for other languages or libraries. 99% of use-cases will use the lunatic-rs for their project anyway, so that's not a big issue, it's just maybe something can easily be overlooked and cause some issues down the road. But maybe that's a premature optimisation, I'm fine with both solutions. I do however see value in you wanting to run your own allocator if you're tight on memory or use a very large buffer that you'd like to re-use. Other than that it makes the "magic" of just having an alloc function more understandable imo

**Comment by bkolobara** [comment]
> I would like this approach as well, but the issue with the sql crate for example is that it doesn't have a dependency on lunatic-rs and this way we'd avoid this dependency.

This is true, but you are also running into issues on the other side of the spectrum. If 2 packages export an allocator that has the same name the linking will fail. Name collision on exports is still an unresolved issue for us and if we encourage people to export more stuff with the name "alloc", this might become somewhat of an issue.

I would prefer just to have a `lunatic-sys`? crate that is minimalistic, exports all this well known stuff and is used as a dependency everywhere (lunatic-rs, sqlite, ...).

> I do however see value in you wanting to run your own allocator if you're tight on memory or use a very large buffer that you'd like to re-use.

Would this work though? A custom allocator, like an arena allocator would need some additional information (pointer to the arena). The host would need to be aware of this and use a different signature providing more data to the allocator. You can always change the global allocator in Rust, but using a specific one just for SQLite is rather difficult in general.

**Comment by SquattingSocrates** [comment]
> I would prefer just to have a `lunatic-sys`? crate that is minimalistic, exports all this well known stuff and is used as a dependency everywhere (lunatic-rs, sqlite, ...).

Yeah, I can do that, no problem. Should this be a separate git repository or do I just create a sub-crate here in the vm and set the target to wasm32?


> Would this work though? 

I don't see why not. There's probably not a lot of use-cases I can currently think of that would require this approach, but I think it's definitely doable with the current api :)

**Comment by bkolobara** [comment]
> Yeah, I can do that, no problem. Should this be a separate git repository or do I just create a sub-crate here in the vm and set the target to wasm32?

No idea. Should probably go under the `lunatic-rs` repo instead as it is guest exclusive?

I'm also not completely sold on the unifying guest/host API approach by using a bunch of `#[cfg(target_arch = "wasm32")]` annotations everywhere :). It makes the code a bit harder to mentally process for me.

**Comment by SquattingSocrates** [comment]
I could split the code into separate files for separate targets. The thing is though that we need to use the same structs and we can achieve that by either cloning the structs to other projects or to export the structs from a single crate.

**Comment by bkolobara** [comment]
I like the splitting into separate files idea for not shared code. The mixing inside one file stuff makes is a bit hard to follow.

**Comment by SquattingSocrates** [comment]
> I like the splitting into separate files idea for not shared code. The mixing inside one file stuff makes is a bit hard to follow.

Alright, I'll do that then.

So, just to sum all the fixes and improvements ahead of me:

- [ ] move wasm code and vm code into separate files
- [ ] go back to using hardcoded name for guest alloc function
- [ ] create lunatic-sys crate and use it as dependency in lunatic-sql and lunatic-rs
- [ ] Change signature of host functions to write the length of the result to an opaque pointer instead of into a u64
- [ ] Test how the implementation handles multiple connections reading and writing to the DB
- [ ] Make sure that only allowed paths are used for creating the DB file

I would probably not try to build an database-agnostic solution just yet, but if you want I could just keep working on that then.

**Review by bkolobara — CHANGES_REQUESTED** [review]
Looks good 😄! I just requested a few cosmetic changes.

File: crates/lunatic-common-api/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SquattingSocrates) [reply crates/lunatic-common-api/src/lib.rs:18]: Trapping should suffice for now I guess. It's either an untrusted program trying to use more memory than it should or a memory limit was not properly set, which means it could be a bug or a miscalculation. Either way, crashing and trapping is probably the most straightforward option.

**Review by bkolobara — APPROVED** [review]
